packages/test-utils/src/reporter.ts

```diff
@@ -920,44 +919,69 @@ class ReplayReporter<TRecordingMetadata extends UnstructuredMetadata = Unstructu
 
       let uploadableResults = this._uploadableResults.get(result.testId);
       if (!uploadableResults) {
-        uploadableResults = [];
+        uploadableResults = {
+          executions: {},
+          uploadedStatuses: {
+            passed: false,
+            failed: false,
+          },
+        };
         this._uploadableResults.set(result.testId, uploadableResults);
       }
-      uploadableResults.push(result);
+      let executions = uploadableResults.executions[result.executionGroupId];
+      if (!executions) {
+        executions = [];
+        uploadableResults.executions[result.executionGroupId] = executions;
+      }
+      executions.push(result);
 
       if (result.result === "passed" || result.attempt >= result.maxAttempts) {
-        this._enqueueUploads(uploadableResults);
+        this._enqueueUploads(uploadableResults, result.executionGroupId);
       }
     }
   }
 
-  private _enqueueUploads(results: UploadableTestRunResult<TRecordingMetadata>[]) {
-    const latestResult = results[results.length - 1];
-    assert(!!latestResult, "Expected at least one result in the list");
-
-    let toUpload: typeof results | undefined;
+  private _enqueueUploads(
+    result: UploadableTestResult<TRecordingMetadata>,
+    executionGroupId: string
+  ) {
+    const executions = result.executions[executionGroupId];
+    const latestExecution = last(executions);
+    assert(!!latestExecution, "Expected at least one execution in the list");
 
+    let toUpload: typeof executions | undefined;
     switch (this._uploadStatusThreshold) {
       case "all":
-        toUpload = this._minimizeUploads ? [latestResult] : results;
+        toUpload = this._minimizeUploads
+          ? this._getPotentiallyFlakyUploads(result, executions)
+          : executions;
         break;
       case "failed-and-flaky":
-        // retries can be disables so we need to always check if the latest result is not passed
+        // retries can be disabled so we need to always check if the latest execution is not passed
         // with retries enabled we know that we only have to upload when there was more than one attempt at the test
         // a single passed attempt can be safely ignored, multiple attempts mean that the test is flaky or failing
-        if (latestResult.result !== "passed" || results.length > 1) {
-          if (!this._minimizeUploads) {
-            toUpload = results;
-          } else if (results.length > 1) {
-            toUpload = [results[0], latestResult];
-          } else {
-            toUpload = [latestResult];
-          }
+        // if there is no failed execution then this test is not flaky or failing
+        if (
+          latestExecution.result !== "passed" ||
+          executions.length > 1 ||
+          (result.uploadedStatuses.failed && !result.uploadedStatuses.passed)
```

Comment:
This looks super funky.

Reply:
Like is this really supposed to be in the `||` chain? I don't really understand the `uploadedStatuses` data structure (I think it could use some inline comments at least) but my intuition is that maybe this is more of an `&&` conditional, else it seems like we would get false positives here?